packages/lodestar/src/sync/backfill/backfill.ts

```diff
@@ -0,0 +1,336 @@
+  [BackfillSyncStatus.syncing]: 1,
+  [BackfillSyncStatus.completed]: 2,
+};
+
+type BackfillSyncEvents = {
+  // slot param is oldest slot synced
+  [BackfillSyncEvent.completed]: (oldestSlotSynced: Slot) => void;
+};
+
+type BackfillSyncEmitter = StrictEventEmitter<EventEmitter, BackfillSyncEvents>;
+
+export class BackfillSync extends (EventEmitter as {new (): BackfillSyncEmitter}) {
+  lastBackSyncedSlot?: Slot | null = null;
+
+  private readonly chain: IBeaconChain;
+  private readonly network: INetwork;
+  private readonly db: IBeaconDb;
+  private readonly config: IBeaconConfig;
+  private readonly logger: ILogger;
+  private readonly metrics: IMetrics | null;
+  private wsCheckpoint?: phase0.Checkpoint | null;
+  private opts: BackfillSyncOpts;
+
+  // At any given point, we should have either the anchorBlock, or anchorBlockRoot, the reversed head of the backfill sync, from where we need to backfill
+  private anchorBlock: allForks.SignedBeaconBlock | null = null;
+  private anchorBlockRoot?: Root | null = null;
+  private backfillStartFromSlot?: Slot;
+
+  private processor = new ItTrigger();
+  private peers = new PeerSet();
+  private status: BackfillSyncStatus = BackfillSyncStatus.pending;
+
+  constructor(modules: BackfillSyncModules, opts?: BackfillSyncOpts) {
+    super();
+    this.chain = modules.chain;
+    this.network = modules.network;
+    this.db = modules.db;
+    this.config = modules.config;
+    this.logger = modules.logger;
+    this.metrics = modules.metrics;
+    this.wsCheckpoint = modules.wsCheckpoint;
+    this.opts = opts ?? {batchSize: BATCH_SIZE};
+    this.network.events.on(NetworkEvent.peerConnected, this.addPeer);
+    this.network.events.on(NetworkEvent.peerDisconnected, this.removePeer);
+
+    this.sync()
+      .then((oldestSlotSynced) => {
+        this.emit(BackfillSyncEvent.completed, oldestSlotSynced);
+        this.logger.info("BackfillSync completed", {oldestSlotSynced});
+        this.status = BackfillSyncStatus.completed;
+        // Sync completed, unsubscribe listeners and don't run the processor again
+        // Backfill is never necessary again until the node shuts down
+        this.close();
+      })
+      .catch((e) => {
+        if (e instanceof ErrorAborted) {
+          return; // Ignore
+        }
+        this.logger.error("BackfillSync processor error", e);
+        this.status = BackfillSyncStatus.completed;
+        this.close();
+      });
+    if (this.metrics) {
+      this.metrics.backfillSync.status.addCollect(() => this.metrics?.backfillSync.status.set(syncStatus[this.status]));
+    }
+  }
+
+  /** Throw / return all AsyncGenerators */
+  close(): void {
+    this.network.events.off(NetworkEvent.peerConnected, this.addPeer);
+    this.network.events.off(NetworkEvent.peerDisconnected, this.removePeer);
+    this.processor.end(new ErrorAborted("BackfillSync"));
+  }
+
+  /** Returns oldestSlotSynced */
+  private async sync(): Promise<Slot> {
```

Comment:
This function is 134 lines long and very hard to follow, please
- Add code comments throughout the code blocks explaining what is going on and how this is intended to work
- If possible try to modularize the code with good abstractions that help the above point self-documenting the code
- As bonus (if applicable) you can add a diagram or summary of the entire flow at the top of the function